File: individual-name-matching/python-examples/generate-match-report.py
import json
import urllib.parse
import urllib.request

# Replace this with your API key from https://www.interzoid.com/manage-api-account
API_KEY = "YOUR_API_KEY_HERE"

# Input file containing one full name per line
INPUT_FILE_NAME = "sample-input-file.txt"


def call_fullname_match_api(full_name: str) -> str:
    """
    Calls Interzoid's getfullnamematch API for a single full name
    and returns the similarity key (SimKey) as a string.
    Returns an empty string if there is an error or no SimKey.
    """

    # URL-encode the full name so it is safe to send as a parameter
    name_param = urllib.parse.quote(full_name)

    api_url = (
        "https://api.interzoid.com/getfullnamematch"
        f"?license={API_KEY}"
        f"&fullname={name_param}"
    )

    try:
        with urllib.request.urlopen(api_url) as response:
            body = response.read().decode("utf-8")
    except Exception as e:
        print(f"Error calling API for '{full_name}': {e}")
        return ""

    try:
        data = json.loads(body)
    except json.JSONDecodeError:
        print(f"Error parsing JSON for '{full_name}'. Raw response was:\n{body}\n")
        return ""

    # Expected JSON: {"SimKey":"...","Code":"Success","Credits":"..."}
    simkey = data.get("SimKey", "")
    return simkey or ""


def main():
    # Each record holds the original input and its SimKey
    records = []

    # Read the input file line by line
    try:
        with open(INPUT_FILE_NAME, "r", encoding="utf-8") as f:
            for line in f:
                full_name = line.strip()

                # Skip blank lines
                if not full_name:
                    continue

                simkey = call_fullname_match_api(full_name)

                # Skip if no SimKey returned
                if not simkey:
                    continue

                records.append({"input": full_name, "simkey": simkey})
    except OSError as e:
        print(f"Error opening or reading input file: {e}")
        return

    if not records:
        print("No records with similarity keys found.")
        return

    # Sort strictly by SimKey so identical keys are adjacent
    records.sort(key=lambda r: r["simkey"])

    current_key = None
    cluster = []

    def print_cluster(c):
        # Only print clusters with two or more records
        if len(c) < 2:
            return
        for rec in c:
            print(f"{rec['input']},{rec['simkey']}")
        print()  # blank line between clusters

    # Walk through sorted records and build clusters by SimKey
    for rec in records:
        if rec["simkey"] != current_key:
            if cluster:
                print_cluster(cluster)
            current_key = rec["simkey"]
            cluster = [rec]
        else:
            cluster.append(rec)

    # Flush the final cluster
    if cluster:
        print_cluster(cluster)


if __name__ == "__main__":
    main()

File: individual-name-matching/go-examples/generate-match-report.go
// generate-match-report.go
//
// This example reads a list of individual names from a text file,
// generates similarity keys (SimKeys) using Interzoid's Individual Name
// Matching API, then groups and prints names that share the same SimKey.
//
// Each printed line is "Input,SimKey" (two-column CSV style).
// Only clusters with two or more matching names are printed.

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"sort"
)

// Replace this with your API key from https://www.interzoid.com/manage-api-account
const API_KEY = "YOUR_API_KEY_HERE"

// Input file containing one full name per line
const inputFileName = "sample-input-file.txt"

func main() {
	// Record holds one input name and its generated similarity key
	type Record struct {
		Input  string
		SimKey string
	}

	// Response maps the JSON returned by the getfullnamematch API
	type Response struct {
		SimKey  string `json:"SimKey"`
		Code    string `json:"Code"`
		Credits string `json:"Credits"`
	}

	// Open the input file for reading
	file, err := os.Open(inputFileName)
	if err != nil {
		fmt.Println("Error opening input file:", err)
		return
	}
	defer file.Close()

	var records []Record

	scanner := bufio.NewScanner(file)

	// Read each line as a full name and call the API
	for scanner.Scan() {
		name := scanner.Text()

		// Skip blank lines
		if name == "" {
			continue
		}

		// URL-encode the name so it can be safely passed as a parameter
		nameParam := url.QueryEscape(name)

		// Build the getfullnamematch API URL
		apiURL := "https://api.interzoid.com/getfullnamematch?license=" + API_KEY +
			"&fullname=" + nameParam

		// Call the API
		resp, err := http.Get(apiURL)
		if err != nil {
			fmt.Println("Error calling API for:", name, "-", err)
			continue
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			fmt.Println("Error reading API response for:", name, "-", err)
			continue
		}

		var result Response

		// Parse the JSON response into our struct
		if err := json.Unmarshal(body, &result); err != nil {
			fmt.Println("Error parsing JSON for:", name, "-", err)
			fmt.Println("Raw response:", string(body))
			continue
		}

		// If there is no SimKey, skip this record
		if result.SimKey == "" {
			continue
		}

		// Store the name and SimKey in memory for later clustering
		records = append(records, Record{
			Input:  name,
			SimKey: result.SimKey,
		})
	}

	// Check for any scan error from reading the file
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}

	if len(records) == 0 {
		fmt.Println("No records with similarity keys found.")
		return
	}

	//----------------------------------------------------------------------
	// Sort the records strictly by SimKey only.
	// This ensures that all identical SimKeys are adjacent in the slice,
	// making it easy to identify groups of matches.
	//----------------------------------------------------------------------
	sort.Slice(records, func(i, j int) bool {
		return records[i].SimKey < records[j].SimKey
	})

	//----------------------------------------------------------------------
	// Walk through the sorted slice and collect clusters of records that
	// share the same SimKey. Only print clusters that contain two or more
	// records. Each line is printed as "Input,SimKey" and clusters are
	// separated by a blank line.
	//----------------------------------------------------------------------

	currentKey := ""
	var cluster []Record

	// Helper function to print a cluster if it has two or more records
	printCluster := func(c []Record) {
		if len(c) < 2 {
			return
		}
		for _, r := range c {
			fmt.Printf("%s,%s\n", r.Input, r.SimKey)
		}
		fmt.Println() // blank line between clusters
	}

	for _, rec := range records {
		if rec.SimKey != currentKey {
			// SimKey changed: flush the previous cluster (if any)
			if len(cluster) > 0 {
				printCluster(cluster)
			}
			currentKey = rec.SimKey
			cluster = []Record{rec}
		} else {
			// Same SimKey as current cluster, add to it
			cluster = append(cluster, rec)
		}
	}

	// Flush the final cluster at the end
	if len(cluster) > 0 {
		printCluster(cluster)
	}
}
